eth_trie/hexary.nim

```diff
@@ -119,16 +122,219 @@ proc getAux(db: DB, nodeRlp: Rlp, path: NibblesRange): BytesRange =
     if branch.isEmpty:
       return zeroBytesRange
     else:
-      let nextLookup = if branch.isList: branch
-                       else: rlpFromBytes db.get(branch.expectHash.toOpenArray).toRange
+      let nextLookup = branch.getLookup
       return getAux(db, nextLookup, path.slice(1))
   else:
     raise newException(CorruptedTrieError,
-                       "HexaryTrie node with an unexpected numbef or children")
+                       "HexaryTrie node with an unexpected number of children")
 
 proc get*(self: HexaryTrie; key: BytesRange): BytesRange =
   return getAuxByHash(self.db, self.root, initNibbleRange(key))
 
+proc getValuesAux(db: DB, nodeRlp: Rlp, leaves: var seq[BytesRange]) =
+  if not nodeRlp.hasData or nodeRlp.isEmpty:
+    return
+
+  case nodeRlp.listLen
+  of 2:
+    let
+      (isLeaf, _) = nodeRlp.extensionNodeKey
+      value = nodeRlp.listElem(1)
+
+    if isLeaf:
+      leaves.add value.toBytes
+    else:
+      let nextLookup = value.getLookup
+      db.getValuesAux(nextLookup, leaves)
+  of 17:
+    for i in 0 ..< 16:
+      var branch = nodeRlp.listElem(i)
+      if not branch.isEmpty:
+        let nextLookup = branch.getLookup
+        db.getValuesAux(nextLookup, leaves)
+    var lastElem = nodeRlp.listElem(16)
+    if not lastElem.isEmpty:
+      leaves.add lastElem.toBytes
+  else:
+    raise newException(CorruptedTrieError,
+                       "HexaryTrie node with an unexpected number of children")
+
+proc getValues*(self: HexaryTrie): seq[BytesRange] =
+  result = @[]
+  var nodeRlp = rlpFromBytes keyToLocalBytes(self.db, self.root)
+  self.db.getValuesAux(nodeRlp, result)
+
+proc getKeysAux(db: DB, nodeRlp: Rlp, keys: var seq[ByteRange], path: NibblesRange) =
+  if not nodeRlp.hasData or nodeRlp.isEmpty:
+    return
+
+  case nodeRlp.listLen
+  of 2:
+    let
+      (isLeaf, k) = nodeRlp.extensionNodeKey
+      key = path & k
```

Comment:
It should be possible to avoid the allocations here. We can allocate a single `NibleRange` having the maximum possible length and we'll push and pop nibbles from it as we go deeper into the Trie.